Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,217 @@
+## Arm64 Intrinsics
+
+This document is intended to document proposed design decision related to the introduction
+of Arm64 Intrinsics
+
+### Intrinsics in general
+
+Use of intrinsics in general is a CoreCLR design decision to allow low level platform
+specific optimizations.  
+
+At first glance, such a decision seems to violate the fundamental principles of .NET
+code running on any platform.  However, the intent is not for the vast majority of
+apps to use such optimizations.  The intended usage model is to allow library
+developers access to low level functions which enable optimization of key
+functions.  As such the use is expected to be limited, but performance critical.
+
+#### Intrinsic `IsSupported` granularity
+
+The existing implementation is using `IsSupported` to determine whether an existing
+namespace is supported on a given platform.
+
+The granularity is kept at a large scale to have large blocks of functionality controlled
+by a single `IsSupported`.
+
+#### Intrinsic granularity
+
+In general individual intrinsic will be chosen to be fine grained.  These will generally correspond to a single assembly instruction.
+
+#### Intrinsics & Crossgen
+
+For any intrinsic which may not be supported on all variants of a platform.  Crossgen
+Method compilation must be trapped, so that the JIT is forced to generate optimal platform dependent 
+code at runtime.
+
+#### API review
+
+Intrinsics will extend the API of CoreCLR.  They will need to follow standard API review practices.
+
+### Choice of Arm64 naming conventions
+
+#### Namespaces
+
+`ARM32` and `ARM64` will follow similar namespace conventions.
+
++ `System.Runtime.Intrinsics` is used for type definitions useful across multiple platforms
++ `System.Runtime.Intrinsics.Arm` is used for enumerations shared across `ARM32` and `ARM64` platforms
++ `System.Runtime.Intrinsics.Arm.Base` is recommended for functionality included in all `ARM32` and `ARM64` platforms
+  + Do we need this or could we assume `ARMv8` is minimum supported platform?
+  + Should we assume `ARMv7` or `ARMv5` is minimum supported platform?
++ `System.Runtime.Intrinsics.Arm.ARMv8` is used for `ARMv8` added functionality included in both `ARM32` and `ARM64` platforms 
++ `System.Runtime.Intrinsics.Arm.ARMv8.Arm64` is used for `ARMv8` functionality only included in `ARM64` platform
++ `System.Runtime.Intrinsics.Arm.ARMv8_1....` is used for `ARMv8.1` functionality.  
+  + Use similar sub namespace names for only included in `ARM64` platform ...
++ `System.Runtime.Intrinsics.Arm.XXX` is used for optional features `CRC32`, `AES`, `SHA1` ... 
```

Comment:
It seems like most intrinsics are optional features. Even floating-point and Advanced SIMD aren't guaranteed to be exist in Armv8 and certain implementations are given the option to exclude them.

I think this should be reworded to indicate that any feature that is detected via a register flag (such as `ID_AA64ISAR0_EL1` and `ID_AA64ISAR1_EL1`) should go in a `Arm.XXX` namespace where `XXX` correspdonds to the feature name.

This includes things like `DP`, `SM4`, `SM3`, `SHA3`, `RDM`, `Atomic`, `CRC32`, `SHA2`, `SHA1`, `AES`, `DPB`, etc.

Reply:
Also, see `ID_AA64PFR0_EL1` and `ID_AA64PFR1_EL1`, which covers `SVE`, `RAS`, `GIC`, `AdvSIMD`, `FP`, `EL3`, `EL2`, `EL1`, and `EL0` (some of which, like `AdvSIMD` and `FP`, are multi-bit and might dictate multiple features).

Reply:
`ID_ISAR0_EL1` through `ID_ISAR6_EL1` are the corresponding feature registers for AArch32

Reply:
As mentioned above, these features registers are not visible from user space.  I will rework the requirements so that features can be tested independently.  

Also note that the set of features is also constrained.  See for example, `ID_AA64ISAR0_EL1.RDM` 
+ `In ARMv8.0 only permitted value is 0000` 
+ `In ARMv8.1 only permitted value is 0001`

Therefore I still think it is useful to allow set tests for testing.

Depending on how these are exposed by the OS, some may need to be inferred.

Reply:
Most of these are already exposed at the OS level (at least on Linux, not sure about MacOS): https://github.com/torvalds/linux/blob/master/arch/arm64/include/uapi/asm/hwcap.h and https://github.com/torvalds/linux/blob/master/arch/arm/include/uapi/asm/hwcap.h

If there are some that are missing, it might be worthwhile getting them added. I believe that attempting to infer these values is just leaving users open to failure later on.

Reply:
Point taken.

With respect to inference, I was mostly referring to Windows ARM64, which unfortunately, I must assume has some mechanism to expose this information.

Reply:
https://msdn.microsoft.com/en-us/library/windows/desktop/ms724482(v=vs.85).aspx